src/html-purify.js

```diff
@@ -39,51 +41,68 @@ See the accompanying LICENSE file for terms.
         /* jshint validthis: true */
         /* jshint expr: true */
         var parser = this.parser,
-            ch = parser.input[i],
-            attributeName = parser.getAttributeName(),
-            attributeValue = parser.getAttributeValue(),
-            idx = parser.getCurrentTagIndex(),
-            tagName = parser.getCurrentTag(idx),
-            attrValString = '';
+            idx, tagName, attrValString, openedTag;
 
         switch (derivedState.Transitions[prevState][nextState]) {
+            
         case derivedState.TransitionName.WITHIN_DATA:
-	    this.output += ch;
-	    break;
+            this.output += parser.input[i];
+        break;
 
         case derivedState.TransitionName.FROM_TAG_ATTR_TO_DATA:
-            if (prevState === 35 ||
-                prevState === 36 || 
-                prevState === 40) {
-                this.attrVals[attributeName] = attributeValue;
-            }
+            idx = parser.getCurrentTagIndex();
+            tagName = parser.getCurrentTag(idx);
 
-            attrValString = '';
-            for (var key in this.attrVals) {
-                if (contains(whitelist.Attributes, key)) {
-                    attrValString += " " + key;
-                    if (this.attrVals[key].length > 0) {
-                        attrValString += "=" + "\"" + this.attrVals[key] + "\"";
+            if (contains(whitelist.Tags, tagName)) {
+
+                if (idx) {
+                    // add closing tags for any opened ones before closing the current one
+                    while((openedTag = this.openedTags.pop()) && openedTag !== tagName) {
+                        this.output += '</' + openedTag + '>';
                     }
-                }
-                else if (contains(whitelist.HrefAttributes, key)) {
-                    attrValString += " " + key;
-                    if (this.attrVals[key].length > 0) {
-                        attrValString += "=" + "\"" + xssFilters.uriInDoubleQuotedAttr(decodeURI(this.attrVals[key])) + "\"";   
+                    // openedTag is undefined if tagName is never found in all openedTags, no output needed
+                    if (openedTag) {
+                        this.output += '</' + openedTag + '>';
                     }
                 }
-                else if (key === "style") {// TODO: move style to a const
-                    if (this.cssParser.parseCssString(this.attrVals[key])) {
-                         attrValString += " " + key + "=" + "\"" + this.attrVals[key] + "\"";
+                else {
+                    //  - void elements only have a start tag; end tags must not be specified for void elements.
+                    this.hasSelfClosing = this.hasSelfClosing || voidElements[tagName];
```

Comment:
any specific reason why we're always adding a solidus for void elements? spec just says it is okay to have a solidus